src/uct/ze/copy/ze_copy_md.c

```diff
@@ -0,0 +1,350 @@
+    }
+
+    if (mem_attr_p->field_mask & UCT_MD_MEM_ATTR_FIELD_ALLOC_LENGTH) {
+        mem_attr_p->alloc_length = mem_info.alloc_length;
+    }
+
+    if (mem_attr_p->field_mask & UCT_MD_MEM_ATTR_FIELD_DMABUF_FD) {
+        mem_attr_p->dmabuf_fd = dmabuf_fd;
+    }
+
+    if (mem_attr_p->field_mask & UCT_MD_MEM_ATTR_FIELD_DMABUF_OFFSET) {
+        mem_attr_p->dmabuf_offset = (uintptr_t)addr -
+                                    (uintptr_t)mem_info.base_address;
+    }
+    return UCS_OK;
+}
+
+static ucs_status_t
+uct_ze_copy_md_detect_memory_type(uct_md_h md, const void *addr, size_t length,
+                                  ucs_memory_type_t *mem_type_p)
+{
+    uct_md_mem_attr_t mem_attr;
+    ucs_status_t status;
+
+    mem_attr.field_mask = UCT_MD_MEM_ATTR_FIELD_MEM_TYPE;
+    status = uct_ze_copy_md_mem_query(md, addr, length, &mem_attr);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    *mem_type_p = mem_attr.mem_type;
+    return UCS_OK;
+}
+
+static uct_md_ops_t md_ops = {
+    .close              = uct_ze_copy_md_close,
+    .query              = uct_ze_copy_md_query,
+    .mem_alloc          = uct_ze_copy_mem_alloc,
+    .mem_free           = uct_ze_copy_mem_free,
+    .mkey_pack          = uct_ze_copy_mkey_pack,
+    .mem_reg            = uct_ze_copy_mem_reg,
+    .mem_dereg          = uct_ze_copy_mem_dereg,
+    .mem_attach         = ucs_empty_function_return_unsupported,
+    .mem_query          = uct_ze_copy_md_mem_query,
+    .detect_memory_type = uct_ze_copy_md_detect_memory_type,
+};
+
+static ucs_status_t
+uct_ze_copy_md_open(uct_component_h component, const char *md_name,
+                    const uct_md_config_t *md_config, uct_md_h *md_p)
+{
+    uct_ze_copy_md_config_t *config = ucs_derived_of(md_config,
+                                                     uct_ze_copy_md_config_t);
+    uct_ze_copy_md_t *md;
+    ze_driver_handle_t ze_driver;
+    ze_context_desc_t context_desc = {};
+    ze_result_t ret;
+
+    ze_driver = uct_ze_base_get_driver();
+    if (ze_driver == NULL) {
+        return UCS_ERR_NO_DEVICE;
+    }
+
+    md = ucs_malloc(sizeof(uct_ze_copy_md_t), "uct_ze_copy_md_t");
+    if (NULL == md) {
+        ucs_error("Failed to allocate memory for uct_ze_copy_md_t");
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    md->ze_device = uct_ze_base_get_device(config->device_ordinal);
+    if (md->ze_device == NULL) {
+        ucs_error("Failed to get device at ordial %d", config->device_ordinal);
+        free(md);
+        return UCS_ERR_NO_DEVICE;
+    }
+
+    ret = zeContextCreate(ze_driver, &context_desc, &md->ze_context);
+    if (ret != ZE_RESULT_SUCCESS) {
+        ucs_error("zeContextCreate failed with error %x", ret);
+        free(md);
+        return UCS_ERR_NO_DEVICE;
+    }
+
+    md->super.ops       = &md_ops;
+    md->super.component = &uct_ze_copy_component;
+
+    *md_p = (uct_md_h)md;
+    return UCS_OK;
+}
+
+uct_component_t uct_ze_copy_component = {
```

Comment:
can we make it static?

Reply:
This can't be static since it is referenced in `ze_copy_iface.c`.